### PR TITLE
[stable/spring-cloud-data-flow]  Added support for external databases and RabbitMQ-HA

### PR DESCRIPTION
**Pull request by lwj5**
#### What this PR does / why we need it:

Added support for external database usage, and RabbitMQ-HA broker, to allow for highly available set-up.

For more details visit:
https://github.com/spring-cloud/spring-cloud-dataflow/issues/3191

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] title of the PR contains starts with chart name e.g. `[stable/chart]`

### CODE REVIEW

**Comment by lwj5** [comment]
@paulczar @cpanato any comments on the modified chart?

File: stable/spring-cloud-data-flow/templates/broker-secret.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  name: {{ printf "%s-%s" .Release.Name "rabbitmq" | trunc 63 | trimSuffix "-" }}
+  labels:
+    app: {{ template "scdf.name" . }}
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+type: Opaque
+data:
+  rabbitmq-password: {{ .Values.broker.rabbitmq.password | b64enc | quote }}
+{{- end }}
```
Comment (cpanato) [inline stable/spring-cloud-data-flow/templates/broker-secret.yaml]: new line here

File: stable/spring-cloud-data-flow/Chart.yaml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 description: Toolkit for building data processing pipelines.
 name: spring-cloud-data-flow
-version: 2.0.2
+version: 2.0.3
```
Comment (cpanato) [inline stable/spring-cloud-data-flow/Chart.yaml]: please bump to 
```suggestion
version: 2.1.0
```

File: stable/spring-cloud-data-flow/templates/database-secret.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    heritage: "{{ .Release.Service }}"
+type: Opaque
+data:
+  {{- if .Values.database.rootPassword }}
+  mysql-root-password: {{ .Values.database.rootPassword | b64enc | quote }}
+  {{- else if (eq .Values.database.user "root")}}
+  mysql-root-password: {{ .Values.database.password | b64enc | quote }}
+  {{- end }}
+  mysql-password:  {{ .Values.database.password | b64enc | quote }}
+{{- end }}
```
Comment (cpanato) [inline stable/spring-cloud-data-flow/templates/database-secret.yaml]: new line here

**Comment by lwj5** [comment]
@cpanato Thanks for the review, updated with changes.

**Comment by sebastien-prudhomme** [comment]
One general question about that chart: for HA why not using rabbitmq-ha and mariadb charts?

**Comment by lwj5** [comment]
@sebastien-prudhomme I guess you could, especially swapping MySQL for MariaDB. MariaDB chart allows for single and multi deployments and this chart is using maria driver as the default.

However, I am not so sure for RabbitMQ HA. I would think the bundle should be light and cater to as wide a range of users as possible, for those without a HA set-up as well. 

I think there is merit in allowing external DB connections, providing more flexibility and maybe in the future adding support for various DB that SCDF supports.

**Comment by lwj5** [comment]
@sebastien-prudhomme I added rabbitmq-ha alongside rabbitmq, so users can choose. If you would like we can collaborate and try to move to MariaDB by creating a major release.

@cpanato please review and approve. Thank you.

File: stable/spring-cloud-data-flow/README.md [hunk 4]
```diff
@@ -101,6 +108,19 @@ The following tables list the configurable parameters and their default values.
...
 
-| Parameter                  | Description           | Default                   |
-| -------------------------- | --------------------- | ------------------------- |
-| mysql.mysqlDatabase        | MySQL database name   | dataflow
+| Parameter                  | Description                  | Default                   |
+| -------------------------- | ---------------------------- | ------------------------- |
+| mysql.enabled              | Enable deployment of MySQL   | true
+| mysql.mysqlDatabase        | MySQL database name          | dataflow
+
+### External MySQL Configuration (mysql.enabled=false)
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: `mysql.enabled` sounds like simply disabling or enabling mysql but in this case the context is to use external access instead. maybe work in "external" to the property name so it's clear.

File: stable/spring-cloud-data-flow/README.md [hunk 5]
```diff
@@ -101,6 +108,19 @@ The following tables list the configurable parameters and their default values.
...
+
+| Parameter                  | Description                    | Default                   |
+| -------------------------- | ------------------------------ | ------------------------- |
+| database.driver            | Database driver                | org.mariadb.jdbc.Driver
+| database.host              | Database host                  | ""
+| database.port              | Database port                  | 3306
+| database.user              | Database user                  | scdf
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: for an external database it would seem like this would already be defined and custom so it should be an override like host.

File: stable/spring-cloud-data-flow/README.md [hunk 6]
```diff
@@ -101,6 +108,19 @@ The following tables list the configurable parameters and their default values.
...
+| database.password          | Database password              | ""
+| database.database.server   | Database name for SCDF server  | dataflow
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: for an external database it would seem like this would already be defined and custom so it should be an override like host.

File: stable/spring-cloud-data-flow/README.md [hunk 7]
```diff
@@ -101,6 +108,19 @@ The following tables list the configurable parameters and their default values.
...
+| database.database.skipper  | Database name for SCDF skipper | skipper
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: for an external database it would seem like this would already be defined and custom so it should be an override like host.

File: stable/spring-cloud-data-flow/README.md [hunk 8]
```diff
@@ -101,6 +108,19 @@ The following tables list the configurable parameters and their default values.
...
+| Parameter                  | Description                  | Default                   |
+| -------------------------- | ---------------------------- | ------------------------- |
+| mysql.enabled              | Enable deployment of MySQL   | true
+| mysql.mysqlDatabase        | MySQL database name          | dataflow
+
+### External MySQL Configuration (mysql.enabled=false)
+
+| Parameter                  | Description                    | Default                   |
+| -------------------------- | ------------------------------ | ------------------------- |
+| database.driver            | Database driver                | org.mariadb.jdbc.Driver
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: seems like this section should be made generic rather than being bound to mysql specifically -- ie: "External Database Configuration" and the default values be empty and supplied by the user. my thinking was it would feel awkward to change this configuration if i was wanting to say use postgres as my "external database"

File: stable/spring-cloud-data-flow/requirements.yaml [hunk 9]
```diff
@@ -2,10 +2,15 @@ dependencies:
...
   repository: https://kubernetes-charts.storage.googleapis.com/
+  condition: mysql.enabled
 - name:  rabbitmq
   version: 0.6.16
   repository: https://kubernetes-charts.storage.googleapis.com/
   condition: rabbitmq.enabled
+- name:  rabbitmq-ha
+  version: 1.27.1
+  repository: https://kubernetes-charts.storage.googleapis.com/
+  condition: rabbitmq-ha.enabled
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/requirements.yaml:13]: are there any notes to indicate when setting `rabbitmq-ha.enabled` to `true`, `rabbitmq.enabled` should be set to `false`? otherwise id assume both would be deployed given `rabbitmq.enabled` defaults to `true` ?

File: stable/spring-cloud-data-flow/README.md (hunk 6 above)
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: having `database.database.` seems redundant

File: stable/spring-cloud-data-flow/README.md (hunk 7 above)
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: having `database.database.` seems redundant

File: stable/spring-cloud-data-flow/templates/database-secret.yaml [hunk 10]
```diff
@@ -0,0 +1,14 @@
+{{- if not .Values.mysql.enabled }}
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/templates/database-secret.yaml:1]: `mysql` references here would be generified if going the route of "external database" vs. "external mysql", including the secret key

File: stable/spring-cloud-data-flow/templates/server-deployment.yaml [hunk 11]
```diff
@@ -77,8 +77,8 @@ spec:
...
             secretName: {{ printf "%s-%s" .Release.Name "mysql" | trunc 63 | trimSuffix "-" }}
       initContainers:
-      - name: init-mysql-wait
-        image: {{ .Values.mysql.busybox.image }}:{{ .Values.mysql.busybox.tag }}
-        imagePullPolicy: {{ .Values.mysql.imagePullPolicy }}
-        command: ['sh', '-c', 'until nc -w3 -z {{ printf "%s-%s" .Release.Name "mysql" | trunc 63 | trimSuffix "-" }} 3306; do echo waiting for mysql; sleep 3; done;']
+      - name: init-db-wait
+        image: {{ .Values.initContainers.dbWait.image }}:{{ .Values.initContainers.dbWait.tag }}
+        imagePullPolicy: {{ .Values.initContainers.dbWait.imagePullPolicy }}
+        command: ['sh', '-c', 'until nc -w3 -z {{ template "scdf.database.host" . }} {{ template "scdf.database.port" . }}; do echo waiting for mysql; sleep 3; done;']
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/templates/server-deployment.yaml]: maybe "waiting for database" rather than "waiting for mysql" as the external DB could be something else?

File: stable/spring-cloud-data-flow/templates/skipper-deployment.yaml [hunk 12]
```diff
@@ -67,27 +67,36 @@ spec:
...
             secretName: {{ printf "%s-%s" .Release.Name "mysql" | trunc 63 | trimSuffix "-" }}
       initContainers:
-      - name: init-mysql-wait
-        image: {{ .Values.mysql.busybox.image }}:{{ .Values.mysql.busybox.tag }}
-        imagePullPolicy: {{ .Values.mysql.imagePullPolicy }}
-        command: ['sh', '-c', 'until nc -w3 -z {{ printf "%s-%s" .Release.Name "mysql" | trunc 63 | trimSuffix "-" }} 3306; do echo waiting for mysql; sleep 3; done;']
+      - name: init-db-wait
+        image: {{ .Values.initContainers.dbWait.image }}:{{ .Values.initContainers.dbWait.tag }}
+        imagePullPolicy: {{ .Values.initContainers.dbWait.imagePullPolicy }}
+        command: ['sh', '-c', 'until nc -w3 -z {{ template "scdf.database.host" . }} {{ template "scdf.database.port" . }}; do echo waiting for mysql; sleep 3; done;']
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/templates/skipper-deployment.yaml]: maybe "waiting for database" rather than "waiting for mysql" as the external DB could be something else?

File: stable/spring-cloud-data-flow/values.yaml [hunk 13]
```diff
@@ -64,4 +74,17 @@ kafka:
     replicaCount: 1
 
 mysql:
+  enabled: true
   mysqlDatabase: dataflow
+
+## If NOT using the MySQL in this chart (mysql.enabled=false),
+## you must specify the following database details
+database:
+  driver: org.mariadb.jdbc.Driver
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/values.yaml]: see comments in README.md

**Review by chrisjs — CHANGES_REQUESTED** [review]
nice additions, just a few things i noticed

File: stable/spring-cloud-data-flow/README.md [hunk 14]
```diff
@@ -8,7 +8,7 @@ Pipelines consist of [Spring Boot](http://projects.spring.io/spring-boot/) apps,
 This chart will provision a fully functional and fully featured Spring Cloud Data Flow installation
 that can deploy and manage data processing pipelines in the cluster that it is deployed to.
 
-MySQL is used as the data store for Spring Cloud Data Flow state and either RabbitMQ or Kafka is used as the messaging layer for streaming apps to communicate with one another.
+MySQL and RabbitMQ are used as the default data store for Spring Cloud Data Flow state and messaging layer for streaming apps to communicate with one another respectively.
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: i'd rather not call RabbitMQ a "data store" here. can this change be reverted?

File: stable/spring-cloud-data-flow/README.md [hunk 15]
```diff
@@ -30,11 +30,30 @@ If you are using a cluster that does not have a load balancer (like Minikube) th
...
+To change the messaging layer to a highly available (HA) version of RabbitMQ, use the following `set` flags to the `helm` command, for example:
+
+`--set rabbitmq-ha.enabled=true,rabbitmq.enabled=false`
+
+Alternatively, to change the messaging layer to Kafka, use the following `set` flags to the `helm` command, for example:
 
 `--set kafka.enabled=true,rabbitmq.enabled=false`
 
-Only one messaging layer can be used at a given time. If RabbitMQ and Kafka are enabled, both charts will be installed with RabbitMQ being used in the deployment.
+Only one messaging layer can be used at a given time. If two or more messaging layers are enabled, the enabled charts will be installed but only one will be used. The order of preference is given by (preferred) RabbitMQ HA > RabbitMQ > Kafka (least preferred).
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: please remove `The order of preference is given by (preferred) RabbitMQ HA > RabbitMQ > Kafka (least preferred).`. realistically, the selection of middleware is determined by `set` flags and if not provided, defaults from values file which is documented in the README.md

File: stable/spring-cloud-data-flow/README.md [hunk 16]
```diff
@@ -101,6 +127,19 @@ The following tables list the configurable parameters and their default values.
...
+| Parameter                  | Description                  | Default                   |
+| -------------------------- | ---------------------------- | ------------------------- |
+| mysql.enabled              | Enable deployment of MySQL   | true
+| mysql.mysqlDatabase        | MySQL database name          | dataflow
+
+### External Database Configuration (mysql.enabled=false)
+
+| Parameter           | Description                    | Default                   |
+| ------------------- | ------------------------------ | ------------------------- |
+| database.driver     | Database driver                | org.mariadb.jdbc.Driver
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: the default values should be generic/empty - at the least `driver` and `port` are mysql specific

**Review by chrisjs — CHANGES_REQUESTED** [review]

File: stable/spring-cloud-data-flow/README.md [hunk 17]
```diff
@@ -101,6 +127,19 @@ The following tables list the configurable parameters and their default values.
...
+| database.host       | Database host                  | ""
+| database.port       | Database port                  | 3306
+| database.user       | Database user                  | scdf
+| database.password   | Database password              | ""
+| database.server     | Database name for SCDF server  | dataflow
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: the name `database.server` sounds like its as it says -- the "database server". i think this could be clarified by say `database.scdf`, `database.dataflow`, etc as with `database.skipper`

File: stable/spring-cloud-data-flow/README.md (hunk 4 above)
Comment (lwj5) [reply stable/spring-cloud-data-flow/README.md]: I disagree with this, many other charts use the same convention. 

https://github.com/helm/charts/tree/master/stable/wordpress
https://github.com/helm/charts/tree/master/stable/prestashop
https://github.com/helm/charts/tree/master/stable/artifactory

I've updated the readme under the section data store, to let users know. If you are ok with this let me know. I will resolve this conversation.

File: stable/spring-cloud-data-flow/README.md (hunk 14 above)
Comment (lwj5) [reply stable/spring-cloud-data-flow/README.md]: It's respectively, meaning MySQL->data store, RabbitMQ->Messaging layer. To make it clearer, I have made it separate, somewhat just like before.

File: stable/spring-cloud-data-flow/README.md (hunk 15 above)
Comment (lwj5) [reply stable/spring-cloud-data-flow/README.md]: The order of preference is used when two or more messaging layers are enabled. I have made it more explicit in the readme, please let me know if it's ok.

File: stable/spring-cloud-data-flow/README.md (hunk 5 above)
Comment (lwj5) [reply stable/spring-cloud-data-flow/README.md]: I'm not sure what you are saying, could you provide more explanation for this.

File: stable/spring-cloud-data-flow/README.md (hunk 4 above)
Comment (chrisjs) [reply stable/spring-cloud-data-flow/README.md]: not debating if this convention is used to enable or disable charts, that's understood. im debating how its used in the context of this line. in short it would likely be better to make a note about `mysql.enable` needing to be false in the content rather than the header, which looks like it has been done so thats fine

File: stable/spring-cloud-data-flow/README.md [hunk 18]
```diff
@@ -8,7 +8,7 @@ Pipelines consist of [Spring Boot](http://projects.spring.io/spring-boot/) apps,
 This chart will provision a fully functional and fully featured Spring Cloud Data Flow installation
 that can deploy and manage data processing pipelines in the cluster that it is deployed to.
 
-MySQL is used as the data store for Spring Cloud Data Flow state and either RabbitMQ or Kafka is used as the messaging layer for streaming apps to communicate with one another.
+MySQL is used as the default data store for Spring Cloud Data Flow state and RabbitMQ is used as the default messaging layer for streaming apps to communicate with one another.
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: please revert this change, there is no need to add default in both these places.

File: stable/spring-cloud-data-flow/README.md [hunk 19]
```diff
@@ -30,11 +30,30 @@ If you are using a cluster that does not have a load balancer (like Minikube) th
...
+To change the messaging layer to a highly available (HA) version of RabbitMQ, use the following `set` flags to the `helm` command, for example:
+
+`--set rabbitmq-ha.enabled=true,rabbitmq.enabled=false`
+
+Alternatively, to change the messaging layer to Kafka, use the following `set` flags to the `helm` command, for example:
 
 `--set kafka.enabled=true,rabbitmq.enabled=false`
 
-Only one messaging layer can be used at a given time. If RabbitMQ and Kafka are enabled, both charts will be installed with RabbitMQ being used in the deployment.
+Only one messaging layer can be used at a given time. If two or more messaging layers are enabled, the enabled charts will be installed but only one will be used, this preference is given by (preferred) RabbitMQ HA > RabbitMQ > Kafka (least preferred). That is to say, if RabbitMQ and Kafka are enabled, only RabbitMQ will be used.
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: please revert this change - we already indicate that only one messaging layer should be enabled at a time and that if both are enabled (user error), both will be deployed. we don't need to go into the details of whats going to be selected here on configuration error.

File: stable/spring-cloud-data-flow/README.md [hunk 20]
```diff
@@ -30,11 +30,30 @@ If you are using a cluster that does not have a load balancer (like Minikube) th
...
+
+By default, MySQL is deployed with this chart. However, if you wish to use an external database, please use the following `set` flags to the `helm` command to disable MySQL deployment, for example:
+
+`--set mysql.enabled=false`
+
+In addition, you are required to set all fields listed in [External Database Configuration](#external-database-configuration-mysqlenabledfalse).
+
+### Messaging Layer
+
+There are three messaging layer available in this chart:
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: layer -> layers

File: stable/spring-cloud-data-flow/values.yaml [hunk 21]
```diff
@@ -64,4 +74,16 @@ kafka:
     replicaCount: 1
 
 mysql:
+  enabled: true
   mysqlDatabase: dataflow
+
+## If NOT using the MySQL in this chart (mysql.enabled=false),
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/values.yaml]: this should probably be generic, something like "## If you are using an external database,"

**Review by chrisjs — CHANGES_REQUESTED** [review]

**Comment by lwj5** [comment]
@chrisjs Thanks for the review. Updated as suggested.

File: stable/spring-cloud-data-flow/README.md [hunk 22]
```diff
@@ -8,7 +8,7 @@ Pipelines consist of [Spring Boot](http://projects.spring.io/spring-boot/) apps,
 This chart will provision a fully functional and fully featured Spring Cloud Data Flow installation
 that can deploy and manage data processing pipelines in the cluster that it is deployed to.
 
-MySQL is used as the data store for Spring Cloud Data Flow state and either RabbitMQ or Kafka is used as the messaging layer for streaming apps to communicate with one another.
+MySQL is used as the data store for Spring Cloud Data Flow state and RabbitMQ is used as the messaging layer for streaming apps to communicate with one another.
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: per previous comments, please revert this change. either rabbit or kafka can be used as the messaging layer. the original content should remain here..

**Review by chrisjs — CHANGES_REQUESTED** [review]

File: stable/spring-cloud-data-flow/README.md (hunk 22 above)
Comment (lwj5) [reply stable/spring-cloud-data-flow/README.md]: Updated. But I don't quite agree with this. If you are willing to put RabbitMQ or Kafka, then for datastore it should be "Either MySQL, PostgreSQL or MariaDB" or "Either MySQL or external databases as such PostgreSQL or MariaDB".

Comment (chrisjs) [reply stable/spring-cloud-data-flow/README.md]: adding something to the effect of "Either the default MySQL deployment or an external database can be used as the data store ....." would be fine. I'd prefer not adding the specifics: "such as PostgreSQL or MariaDB" as the maintained list of supported databases can be found in the SCDF docs...

File: stable/spring-cloud-data-flow/README.md [hunk 23]
```diff
@@ -101,6 +127,19 @@ The following tables list the configurable parameters and their default values.
...
 
-| Parameter                  | Description           | Default                   |
-| -------------------------- | --------------------- | ------------------------- |
-| mysql.mysqlDatabase        | MySQL database name   | dataflow
+| Parameter                  | Description                  | Default                   |
+| -------------------------- | ---------------------------- | ------------------------- |
+| mysql.enabled              | Enable deployment of MySQL   | true
+| mysql.mysqlDatabase        | MySQL database name          | dataflow
+
+### External Database Configuration (mysql.enabled=false)
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: this still needs to be changed.. it should be `### External Database Configuration`

up higher in the readme we already know this parameter needs to be set and if out of the box support for another DB comes along, this is yet another place to update. please remove, thanks.

File: stable/spring-cloud-data-flow/README.md [hunk 24]
```diff
@@ -30,7 +30,26 @@ If you are using a cluster that does not have a load balancer (like Minikube) th
...
 ```
 
-RabbitMQ is the default messaging layer, Kafka can be enabled instead by using the following `set` flags to the `helm` command, for example:
+### Data Store
+
+By default, MySQL is deployed with this chart. However, if you wish to use an external database, please use the following `set` flags to the `helm` command to disable MySQL deployment, for example:
+
+`--set mysql.enabled=false`
+
+In addition, you are required to set all fields listed in [External Database Configuration](#external-database-configuration-mysqlenabledfalse).
```
Comment (chrisjs) [inline stable/spring-cloud-data-flow/README.md]: `(#external-database-configuration-mysqlenabledfalse)` should be updated when the comment below about `mysql.enabled=false` is removed from the header

**Review by chrisjs — CHANGES_REQUESTED** [review]

**Comment by chrisjs** [comment]
when manually testing your changes, so far i have run into the following:

* issue 1:
when enabling rabbitmq-ha and deploying a simple stream, say `time | log`, the applications fail to connect to rabbit, to start:

```
    Property: spring.rabbitmq.port
    Value: ${MY_RELEASE_RABBITMQ_HA_SERVICE_PORT_AMQP}
    Reason: Failed to convert property value of type 'java.lang.String' to required type 'int' for property 'port'; nested exception is org.springframework.core.convert.ConverterNotFoundException: No converter found capable of converting from type [java.lang.String] to type [int]
```

* issue 2:
when setting `mysql.enabled=false` and providing all required vars from the README, in an attempt to use an external database, both skipper and scdf pods are failing to init, ie:

```
  Normal   Scheduled    3m32s                default-scheduler  Successfully assigned default/my-release-data-flow-server-5c777df6d8-sp2zl to minikube
Warning  FailedMount  85s (x9 over 3m32s)  kubelet, minikube  MountVolume.SetUp failed for volume "mysql" : secret "my-release-mysql" not found
```

```
  Normal   Scheduled    5m7s                 default-scheduler  Successfully assigned default/my-release-data-flow-skipper-6b9577777f-xgdlk to minikube
  Warning  FailedMount  57s (x10 over 5m7s)  kubelet, minikube  MountVolume.SetUp failed for volume "mysql" : secret "my-release-mysql" not found
```

please check and ensure changes are tested, thanks

**Comment by lwj5** [comment]
Resolved both issues.

Issue 1. rabbitmq-ha uses headless service, used service hostname.
Issue 2, updated to use "my-release-database" when using external db.

Tested with
- [x] internal MySQL, adding apps.
- [x] external MySQL, adding apps.
- [x] RabbitMQ `time | log`.
- [x] RabbitMQ HA `time | log`.

**Comment by chrisjs** [comment]
thanks - things are looking better. with that said, can you squash all these commits?

**Comment by lwj5** [comment]
Squashed. Thanks.

**Comment by vladkarasyov** [comment]
How add external database postgresql ?? 
my value variable 

database: 
  driver: org.postgresql.Driver
  scheme: 
  host: ${host}
  port: 5432
  user: scdf
  password: 
  dataflow: dataflow
  skipper: skipper

image

**Comment by lwj5** [comment]
Hi @vladkarasyov 

This is not a support forum, do not post your questions here. Use https://stackoverflow.com/questions/tagged/spring-cloud-dataflow next time.

You have to provide "postgresql" as the scheme
database:
scheme: postgresql
